src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -51,7 +51,13 @@ public static ConnectionService Instance
            
         private readonly Dictionary<string, ConnectionInfo> ownerToConnectionMap = new Dictionary<string, ConnectionInfo>();
 
-        private readonly ConcurrentDictionary<string, CancellationTokenSource> ownerToCancellationTokenSourceMap = new ConcurrentDictionary<string, CancellationTokenSource>();
+        /// <summary>
+        /// A map containing all CancellationTokenSource objects that are associated with a given URI/ConnectionType pair. 
+        /// Entries in this map correspond to DbConnection instances that are in the process of connecting. 
+        /// Tuple is used here becasue of it's predictable hash code behavior. 
+        /// </summary>
+        private readonly ConcurrentDictionary<Tuple<string, ConnectionType>, CancellationTokenSource> CancelTupleToCancellationTokenSourceMap = 
```

Comment:
We should avoid overloading anything in the Contracts folders - these are serializable objects and in the future we might auto-gen the contracts. I'd recommend subclassing that and implementing the equals and hashcode in the subclass instead. That would separate the data interchange objects from stuff that can be used internally and have more complex methods / additional data defined. What do you think?